### PR TITLE
Solved #2556 : Implement API search for storage locations

### LINKED ISSUE #2556: Storage Location Autocomplete search is non-functional
### **_Description_**

The search functionality in the "Assign Storage Location" pop-up is currently broken. When a user attempts to search for a storage unit (eg. a Freezer or Room) to move a sample, typing in the search box yields no results, even when valid storage units exist in the database.

Upon inspecting the code, the handleSearch function in `AutocompleteMode.jsx` is currently empty and marked with a // TODO, meaning no API call is being made to the backend.

Image

### **_Steps to Reproduce_**

1. Log in to OpenELIS.
2. Navigate to Lab > Samples and ensure a sample exists and is "Received."
3. Navigate to Storage > Samples.
4. Select an active sample and click the "Move" (or Assign) button.
5. In the popup, type the name of an existing storage unit (e.g., "Freezer") into the "Search for location..." box.

### **_Expected Behavior_**

The application should query the backend for matching storage locations.

A dropdown list should appear showing storage units that match the search query (eg "Freezer A").

### **_Actual Behavior_**

- The dropdown remains empty.
- No feedback is provided to the user.
- No network request is triggered in the browser console.

### **_Screenshots_**

Image

### PR DESCRIPTION
**Pull request by shreyaakalra**
Fixes #2556

**Description**
Implemented the missing API integration in `AutocompleteMode.jsx`.
This fixes a bug where the "Assign Storage Location" search box was non-functional and returned no results even when valid storage units existed.

**Changes**
- Connected the `handleSearch` function to the `/rest/storage/devices/search` endpoint.
- Mapped the API response to the Carbon ComboBox format.
- Added input validation to prevent empty API calls.

**Testing**
1. Navigate to **Storage > Samples**.
2. Select a sample and click **Assign > Manage Location**.
3. Type the name of an existing storage device (eg "Freezer").
4. Confirmed that the dropdown now populates with matching results.

**Screenshots**

**Before (Issue #2556)** 
Screenshot 2026-01-15 at 4 55 32 AM

**After (Fix)** 
Screenshot 2026-01-15 at 4 39 25 AM

### CODE REVIEW

File: frontend/src/components/storage/StorageLocationSelector/AutocompleteMode.jsx [hunk 1]
```diff
@@ -11,8 +11,28 @@ const AutocompleteMode = ({ onLocationChange }) => {
...
   const handleSearch = (inputValue) => {
-    // TODO: Implement search API call
-    // For now, empty implementation
+
+    if (!inputValue) {
+      setSearchResults([]);
+      return;
+    }
+
+    fetch(`/rest/storage/devices/search?q=${inputValue}`)
```
Comment (vsvishalsharma) [inline frontend/src/components/storage/StorageLocationSelector/AutocompleteMode.jsx]: This fires an API request on every keystroke. If user types "Laboratory", that's 10 API calls! Consider Debouncing

Comment (vsvishalsharma) [inline frontend/src/components/storage/StorageLocationSelector/AutocompleteMode.jsx]: User input is directly interpolated into URL without encoding
A better opton would be 

`fetch(`/rest/storage/devices/search?q=${encodeURIComponent(inputValue)}`)`

File: frontend/src/components/storage/StorageLocationSelector/AutocompleteMode.jsx [hunk 2]
```diff
@@ -11,8 +11,28 @@ const AutocompleteMode = ({ onLocationChange }) => {
...
+        const formattedResults = data.map((item) => ({
+          ...item,
+          hierarchicalPath: item.hierarchicalPath || item.name,
+        }));
+        
+        setSearchResults(formattedResults);
+      })
+      
+      .catch((error) => {
+        console.error("Error fetching storage locations:", error);
```
Comment (vsvishalsharma) [inline frontend/src/components/storage/StorageLocationSelector/AutocompleteMode.jsx]: console error shouldn't be neccessary

**Review by vsvishalsharma — CHANGES_REQUESTED** [review]

**Comment by shreyaakalra** [comment]
@vsvishalsharma I've made all the changes (added debouncing and encoded the URL)

**Comment by shreyaakalra** [comment]
Hi @vsvishalsharma @pmanko @mozzy11 I have successfully rebased this branch to clean up the commit history and remove unrelated commits. All automated checks are now passing. Ready for your final review!

**Review by vsvishalsharma — APPROVED** [review]

**Comment by shreyaakalra** [comment]
@pmanko I have updated this PR to address the architectural and security concerns raised by the copilot review.

I have also added test coverage to verify debouncing,empty input handling, and API error scenarios, all of which are passing locally.

It's ready for re-review.

**Review by pmanko — DISMISSED** [review]

**Comment by pmanko** [comment]
seems like a few cleanup suggestions, and looks good

**Comment by shreyaakalra** [comment]
@pmanko, reviewed all the copilot suggestions and made the necessary changes.

**Comment by shreyaakalra** [comment]
@mozzy11 can you kindly review this.

**Comment by mozzy11** [comment]
Thank you for this contribution, and apologies that it sat unreviewed for so long.

We are doing a cleanup of the PR queue: this PR is several months old and has unresolved merge conflicts with develop. The codebase has changed substantially in that time (the frontend moved to Vite and is being migrated to TypeScript, the test infrastructure was rewritten, and several modules — test catalog, analyzers, EQA, storage — were reworked), so a mechanical rebase is unlikely to succeed.

Closing as stale per our queue policy. **If the problem you fixed still exists on current develop, please open a fresh PR against the latest code — we are actively reviewing now and it will get prompt attention.** Thanks again for contributing to OpenELIS!
